src/addonloader/luaaddonstate.cpp

```diff
@@ -128,46 +128,107 @@ std::tuple<> LuaAddonState::logImpl(const char *msg) {
-            state_->lua_pushinteger(keyEvent.key().sym());
-            state_->lua_pushinteger(keyEvent.key().states());
-            state_->lua_pushboolean(keyEvent.isRelease());
-            int rv = state_->lua_pcallk(3, 1, 0, 0, nullptr);
+            if (pushArguments) {
+                argc = pushArguments(state_, event);
+            }
+            int rv = state_->lua_pcallk(argc, 1, 0, 0, nullptr);
             if (rv != 0) {
                 LuaPError(rv, "lua_pcall() failed");
                 LuaPrintError(*this);
             } else if (state_->lua_gettop() >= 1) {
-                auto b = state_->lua_toboolean(-1);
-                if (b) {
-                    keyEvent.filterAndAccept();
+                if (handleReturnValue) {
+                    handleReturnValue(state_, event);
                 }
             }
             state_->pop(state_->lua_gettop());
         });
+}
+
+std::tuple<int> LuaAddonState::watchEventImpl(int eventType,
+                                              const char *function) {
+    int newId = currentId_ + 1;
+    std::unique_ptr<HandlerTableEntry<EventHandler>> handler = nullptr;
+
+    switch (static_cast<EventType>(eventType)) {
+    case EventType::InputContextCreated:
+        handler = watchEvent<InputContextCreatedEvent>(
+            EventType::InputContextCreated, newId);
+        break;
+    case EventType::InputContextDestroyed:
+        handler = watchEvent<InputContextDestroyedEvent>(
+            EventType::InputContextDestroyed, newId);
+        break;
+    case EventType::InputContextFocusIn:
+        handler =
+            watchEvent<FocusInEvent>(EventType::InputContextFocusIn, newId);
+        break;
+    case EventType::InputContextFocusOut:
+        handler =
+            watchEvent<FocusOutEvent>(EventType::InputContextFocusOut, newId);
+        break;
+    case EventType::InputContextKeyEvent:
+        handler = watchEvent<KeyEvent>(
+            EventType::InputContextKeyEvent, newId,
+            [](std::unique_ptr<LuaState> &state, KeyEvent &event_) -> int {
+                state->lua_pushinteger(event_.key().sym());
+                state->lua_pushinteger(event_.key().states());
+                state->lua_pushboolean(event_.isRelease());
+                return 3;
+            },
+            [](std::unique_ptr<LuaState> &state, KeyEvent &event_) {
+                auto b = state->lua_toboolean(-1);
+                if (b) {
+                    event_.filterAndAccept();
+                }
+            });
+        break;
+    case EventType::InputContextCommitString:
+        handler = watchEvent<CommitStringEvent>(
+            EventType::InputContextCommitString, newId,
+            [](std::unique_ptr<LuaState> &state,
+               CommitStringEvent &event_) -> int {
+                state->lua_pushstring(event_.text().c_str());
+                return 1;
+            });
+        break;
+    case EventType::InputContextInputMethodActivated:
```

Comment:
This can be merged with the event below. By using InputMethodNotificationEvent as base class.

Reply:
Updated.